File: src/main/java/org/occidere/githubnotifier/service/LineMessengerServiceImpl.java
package org.occidere.githubnotifier.service;

import com.linecorp.bot.client.LineMessagingClient;
import com.linecorp.bot.model.PushMessage;
import com.linecorp.bot.model.message.TextMessage;
import com.linecorp.bot.model.response.BotApiResponse;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.StringUtils;
import org.occidere.githubnotifier.vo.GithubFollower;
import org.occidere.githubnotifier.vo.GithubFollowerDiff;
import org.occidere.githubnotifier.vo.GithubRepositoryDiff;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

/**
 * @author occidere
 * @Blog: https://occidere.blog.me
 * @Github: https://github.com/occidere
 * @since 2020-04-14
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class LineMessengerServiceImpl implements LineMessengerService {

    private final LineMessagingClient lineMessagingClient;

    @Value("${line.bot.id}")
    private String botId;

    @Override
    public void sendFollowerMessageIfExist(GithubFollowerDiff githubFollowerDiff) {
        sendMessageIfExist(createFollowerMessage(githubFollowerDiff));
    }

    @Override
    public void sendRepositoryMessageIfExist(GithubRepositoryDiff githubRepositoryDiff) {
        if (githubRepositoryDiff.isNewChanged() || githubRepositoryDiff.isDeletedChanged()) {
            sendMessageIfExist(createRepositoryMessage(githubRepositoryDiff));
        }
    }

    private String createFollowerMessage(GithubFollowerDiff githubFollowerDiff) {
        String msg = "";

        final List<GithubFollower> newFollowers = githubFollowerDiff.getNewFollowers();
        final List<GithubFollower> deletedFollowers = githubFollowerDiff.getDeletedFollowers();

        if (CollectionUtils.isNotEmpty(newFollowers)) {
            msg += String.format("[%d of new followers]\n%s\n",
                newFollowers.size(),
                newFollowers.stream()
                    .map(GithubFollower::getLogin)
                    .sorted()
                    .collect(Collectors.joining("\n"))
            );
        }

        if (CollectionUtils.isNotEmpty(deletedFollowers)) {
            msg += String.format("[%d of deleted followers]\n%s\n",
                deletedFollowers.size(),
                deletedFollowers.stream()
                    .map(GithubFollower::getLogin)
                    .sorted()
                    .collect(Collectors.joining("\n"))
            );
        }

        return msg;
    }

    private String createRepositoryMessage(GithubRepositoryDiff diff) {
        final int indent = 4;
        StringBuilder msg = new StringBuilder("[" + diff.getName() + "]\n");

        if (diff.isNewChanged()) {
            msg.append("[NEW]\n")
                    .append(CollectionUtils.isEmpty(diff.getNewStargazersLogin()) ? "" :
                            "- STAR\n" + toSortedListString(diff.getNewStargazersLogin(), indent) + "\n")
                    .append(CollectionUtils.isEmpty(diff.getNewForksLogin()) ? "" :
                            "- FORK\n" + toSortedListString(diff.getNewForksLogin(), indent) + "\n")
                    .append(CollectionUtils.isEmpty(diff.getNewWatchersLogin()) ? "" :
                            "- WATCH\n" + toSortedListString(diff.getNewWatchersLogin(), indent) + "\n");
        }

        if (diff.isDeletedChanged()) {
            msg.append("[DELETED]\n")
                    .append(CollectionUtils.isEmpty(diff.getDeletedStargazersLogin()) ? "" :
                            "- STAR\n" + toSortedListString(diff.getDeletedStargazersLogin(), indent) + "\n")
                    .append(CollectionUtils.isEmpty(diff.getDeletedForksLogin()) ? "" :
                            "- FORK\n" + toSortedListString(diff.getDeletedForksLogin(), indent) + "\n")
                    .append(CollectionUtils.isEmpty(diff.getDeletedWatchersLogin()) ? "" :
                            "- WATCH\n" + toSortedListString(diff.getDeletedWatchersLogin(), indent) + "\n");
        }

        return msg.toString();
    }

    private void sendMessageIfExist(String msg) {
        if (StringUtils.isNotBlank(msg)) {
            try {
                BotApiResponse response = lineMessagingClient.pushMessage(new PushMessage(botId, new TextMessage(msg))).get();
                log.info("{}", response);
            } catch (InterruptedException | ExecutionException e) {
                log.error("Fail to send push message!", e);
            }
        }
    }

    private static String toSortedListString(final List<String> logins, final int indent) {
        final String prefix = " ".repeat(Math.max(0, indent)) + "- ";
        return logins.stream()
            .sorted()
            .map(login -> prefix + login + "\n")
            .collect(Collectors.joining());
    }
}

File: src/main/scala/org/occidere/githubnotifier/service/LineMessengerServiceImpl.scala
package org.occidere.githubnotifier.service

import java.util.Properties

import com.linecorp.bot.client.LineMessagingClient
import com.linecorp.bot.model.PushMessage
import com.linecorp.bot.model.message.TextMessage
import com.linecorp.bot.model.response.BotApiResponse
import lombok.Value
import org.apache.commons.lang3.StringUtils
import org.occidere.githubnotifier.service.LineMessengerServiceImpl._
import org.occidere.githubnotifier.vo.{GithubFollowerDiff, GithubRepositoryDiff}
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.io.ClassPathResource
import org.springframework.core.io.support.PropertiesLoaderUtils
import org.springframework.stereotype.Service

import scala.util.{Failure, Success, Try}

/**
 * @author occidere
 * @Blog: https://blog.naver.com/occidere
 * @Github: https://github.com/occidere
 * @since 2020. 08. 15.
 */
object LineMessengerServiceImpl {
  private val log = LoggerFactory.getLogger(classOf[LineMessengerServiceImpl])

  implicit class IndentUtils(it: Iterable[String]) {
    def toMessage(`type`: String, indent: Int = 4): String = {
      if (it.nonEmpty) {
        val prefix = " ".repeat(Math.max(0, indent)) + "- "
        s"- ${`type`}\n${it.toList.sorted.map(x => prefix + x + "\n").mkString("")}"
      } else ""
    }
  }

}

@Service
@Autowired
class LineMessengerServiceImpl(lineMessagingClient: LineMessagingClient) extends LineMessengerService {
  private val props: Properties = PropertiesLoaderUtils.loadProperties(new ClassPathResource("/application.properties"))

//  @Value("${line.bot.id}")
  private val botId: String = props.getProperty("line.bot.id", "")

  override def sendFollowerMessageIfExist(githubFollowerDiff: GithubFollowerDiff): Unit = sendMessageIfExist(createFollowerMessage(githubFollowerDiff))

  override def sendRepositoryMessageIfExist(githubRepositoryDiff: GithubRepositoryDiff): Unit = sendMessageIfExist(createRepositoryMessage(githubRepositoryDiff))

  private def createFollowerMessage(githubFollowerDiff: GithubFollowerDiff): String = {
    val msg: StringBuilder = new StringBuilder()

    val newFollowers = githubFollowerDiff.newFollowers
    if (newFollowers.nonEmpty) msg ++=
      s"""[${newFollowers.size} of new followers]
         |${newFollowers.map(_.login).toList.sorted.mkString("\n")}
         |""".stripMargin

    val deletedFollowers = githubFollowerDiff.deletedFollowers
    if (deletedFollowers.nonEmpty) msg ++=
      s"""[${deletedFollowers.size} of deleted followers]
         |${deletedFollowers.map(_.login).toList.sorted.mkString("\n")}
         |""".stripMargin

    msg.result()
  }

  private def createRepositoryMessage(diff: GithubRepositoryDiff): String = {
    val msg: StringBuilder = new StringBuilder()

    if (diff.hasNewChanged) msg ++=
      s"""[NEW]
         |${diff.newStargazersLogin.toMessage("- STAR")}
         |${diff.newForksLogin.toMessage("- FORK")}
         |${diff.newWatchersLogin.toMessage("- WATCH")}
         |""".stripMargin

    if (diff.hasDeletedChanged) msg ++=
      s"""[DELETED]
         |${diff.deletedStargazersLogin.toMessage("- STAR")}
         |${diff.deletedForksLogin.toMessage("- FORK")}
         |${diff.deletedWatchersLogin.toMessage("- WATCH")}
         |""".stripMargin

    msg.result()
  }

  def sendMessageIfExist(msg: String): Unit = if (StringUtils.isNotBlank(msg))
    Try {
      lineMessagingClient.pushMessage(new PushMessage(botId, new TextMessage(msg))).get()
    } match {
      case Success(response: BotApiResponse) => log.info(s"$response")
      case Failure(e) => log.error("Fail to send push message!", e)
    }
}